COMPACT_DENYLIST_FORMAT.md

```diff
@@ -0,0 +1,486 @@
+content-addressing and efficient transmission over IPFS. However, the
+protocols, subscriptions and list-sharing approaches are rightfully beyond
+this spec.
+
+Beyond all of that, we put emphasis in making our format easily editable by
+users and facilitating integrations using scripts and with other applications
+(unrelated to the implementation of the parsing/blocking inside IPFS). We
+conciously avoid JSON and other machine formats and opt for text and for
+space-delimited items in a grep/sed/cut-friendly way. For example, we expect
+that the following should just work accross implementations for adding and
+blocking something new:
+
+```
+echo /ipfs/QmecDgNqCRirkc3Cjz9eoRBNwXGckJ9WvTdmY16HP88768 >> ~/.config/ipfs/custom.deny
+```
+
+We conciously avoid defining any other API other than expecting
+implementations to honor blocking what is on the denylist and act accordingly
+when it is updated. CLI commands or API endpoint to modify list items etc. are
+outside the scope of this spec. Implementations how much information to
+provide to users when a request for an IPFS object is blocked.
+
+As a last note, if we take Kubo and the go-ipfs stack as the reference IPFS
+implementation, we expect the blocking-layer (that is, the introduction of the
+logic that decides whether an item is blocked or not), to happen cleanly at
+the `NameSystem`, `path.Resolver` and `BlockService` interfaces (IPNS, IPFS
+Path and CID blocks respectively).
+
+This specification corresponds to V1 of the compact list format. We have
+limited the number of features and extensions to a minimum to start working
+with, leaving some ideas on the table and the door open to develop the format
+in future versions.
+
+## Specification
+
+### Denylist file extension, locations and order
+
+While not pertaining to the denylist format itself, we introduce the following conventions about denylist files when they are stored in the local filesystem:
+
+- Denylist files are named with the extension `.deny`.
+- Implementations should look in `/etc/ipfs/denylists/` and
+  `$XDG_CONFIG_HOME/ipfs/denylists/` for denylist files.
+- Denylist files are processed in alphabetical order so that rules from later
+  denylists override rules from earlier denylists on conflict.
+
+### Denylist format
+
+#### Summary
+
+The following example showcases the features and syntax of a compact denylist:
+
+```
+version: 1
+name: IPFSCorp blocking list
+description: A collection of bad things we have found in the universe
+author: abuse-ipfscorp@example.com
+hints:
+  hint: value
+  hint2: value2
+---
+# Blocking by CID - blocks wrapped multihash.
+# Does not block subpaths.
+/ipfs/bafybeihvvulpp4evxj7x7armbqcyg6uezzuig6jp3lktpbovlqfkuqeuoq
+
+# Block all subpaths
+/ipfs/QmdWFA9FL52hx3j9EJZPQP1ZUH8Ygi5tLCX2cRDs6knSf8/*
+
+# Block some subpaths (equivalent rules)
+/ipfs/Qmah2YDTfrox4watLCr3YgKyBwvjq8FJZEFdWY6WtJ3Xt2/test*
+/ipfs/QmTuvSQbEDR3sarFAN9kAeXBpiBCyYYNxdxciazBba11eC/test/*
```

Comment:
Say you are running a gateway only, perhaps the only place you want to block is at the namesys and resolver levels.

So if you put `/ipfs/cid`  in the denylist, it would still be fine as long as you  are not blocking in the BlockService-level.

Our gateways/kubo need more comprehensive blocking (as, I don't want to "fetch" a blocked CID in the first place), but I can imagine applications which are ok only blocking by path and not by block, and allow to fetch blocks to resolve paths, even if paths to those specific blocks are blocked.

Also note that if "/ipfs/cid" is blocked, we see this issue with subpaths. But if we are talking of "/ipfs/cid/path" and "/ipfs/cid/path/other", the latter might still work, depending on how resolution is implemented. So `/*`  has a point there.

Reply:
Thanks, this makes sense, I think the problem here is that this is useless example of the path blocking:

```
# Block all subpaths
/ipfs/QmdWFA9FL52hx3j9EJZPQP1ZUH8Ygi5tLCX2cRDs6knSf8/*
```


We have better example below, with actual subpath:

```
# Block some subpaths (equivalent rules)
/ipfs/Qmah2YDTfrox4watLCr3YgKyBwvjq8FJZEFdWY6WtJ3Xt2/test*
```

I think we should just remove the `/ipfs/cid/*` example as it is only mudding the water.
`/ipfs/cid/subdpath/*` is way better.

Reply:
I have removed the example and added NOTEs explaning and warning about the behaviour here.